src/sentry/workflow_engine/migrations/0049_remove_orphaned_rule_workflows.py

```diff
@@ -0,0 +1,84 @@
+import logging
+from typing import Any
+
+from django.apps.registry import Apps
+from django.db import migrations, router, transaction
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+
+from sentry.new_migrations.migrations import CheckedMigration
+
+logger = logging.getLogger(__name__)
+
+
+def remove_orphaned_rule_workflows(apps: Apps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    Rule = apps.get_model("sentry", "Rule")
+    AlertRuleWorkflow = apps.get_model("workflow_engine", "AlertRuleWorkflow")
+    DataConditionGroup = apps.get_model("workflow_engine", "DataConditionGroup")
+    Workflow = apps.get_model("workflow_engine", "Workflow")
+    Action = apps.get_model("workflow_engine", "Action")
+
+    def _delete_workflow(workflow: Any) -> bool:
+        action_filters = DataConditionGroup.objects.filter(
+            workflowdataconditiongroup__workflow=workflow
+        )
+
+        actions = Action.objects.filter(
+            dataconditiongroupaction__condition_group__in=action_filters
+        )
+
+        # Delete the actions associated with a workflow, this is not a cascade delete
+        # because we want to create a UI to maintain notification actions separately
+        if actions:
+            actions.delete()
+
+        if action_filters:
+            action_filters.delete()
+
+        if workflow.when_condition_group:
+            workflow.when_condition_group.delete()
+
+        workflow.delete()
+
+        return True
+
+    rule_ids = set(Rule.objects.all().values_list("id", flat=True))
```

Comment:
do we need to make this a set?

Reply:
maybe not? does calling values_list() pull it all into memory though?

Reply:
i suppose it does not need to be a set because it's already unique